File: .claude/hooks/modules/sessionstart/12_inflight_reanchor.py
"""§in-flight re-anchor — surface in-flight work at SessionStart.

If the current session already claimed an item, inject that single item back
into the seat. Otherwise, surface the current in-flight campaign items with the
same short report-back contract used at turn end.
"""
from __future__ import annotations

import json
import pathlib
import re
import sys
import tomllib
from typing import Optional

sys.path.insert(0, str(pathlib.Path(__file__).resolve().parent.parent.parent))
from orchestrator.result import HookResult  # noqa: E402  # pyright: ignore[reportMissingImports]
from lib.astgrep_gate import find_project_root  # noqa: E402  # pyright: ignore[reportMissingImports]

MODULE_NAME = "12_inflight_reanchor"
SESSIONSTART_SOURCES = {"startup", "resume", "clear", "compact"}
LINE_ONE = "§in-flight re-anchor — in_flight at session start:"
GUIDANCE = (
    "If YOURS: note progress via python3 dev/campaigns/manifest.py <path> note <ID> ...; "
    "set-status done only after its verify passes; then reply ONE line: <ID> done — see ledger. "
    "Laws: manifest.py laws"
)
MAX_PAYLOAD_CHARS = 500
MAX_ITEMS = 3
TITLE_LIMIT = 60


def _safe_id(value: object) -> str:
    raw = str(value or "default")
    return re.sub(r"[^a-zA-Z0-9._-]", "_", raw)[:128] or "default"


def _read_json(path: pathlib.Path) -> dict:
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return {}
    return data if isinstance(data, dict) else {}


def _read_toml(path: pathlib.Path) -> dict:
    try:
        data = tomllib.loads(path.read_text(encoding="utf-8"))
    except (OSError, tomllib.TOMLDecodeError):
        return {}
    return data if isinstance(data, dict) else {}


def _pointer_path(root: pathlib.Path, session_id: object) -> pathlib.Path:
    return root / ".claude" / "state" / f"ledger-active-{_safe_id(session_id)}.json"


def _manifest_path(root: pathlib.Path, raw_path: object) -> pathlib.Path | None:
    if not isinstance(raw_path, str) or not raw_path:
        return None
    path = pathlib.Path(raw_path)
    if not path.is_absolute():
        path = root / path
    try:
        return path.resolve(strict=False)
    except OSError:
        return None


def _campaign_inflight_items(campaigns: pathlib.Path) -> list[dict[str, str]]:
    items: list[dict[str, str]] = []
    if not campaigns.is_dir():
        return items

    for manifest in sorted(campaigns.glob("*.toml")):
        doc = _read_toml(manifest)
        for item in doc.get("items", []):
            if item.get("status") != "in_flight":
                continue
            items.append(
                {
                    "id": str(item.get("id", "?")),
                    "stem": manifest.stem,
                    "title": str(item.get("title", ""))[:TITLE_LIMIT],
                }
            )
    return items


def _pointer_item(root: pathlib.Path, session_id: object) -> dict[str, str] | None:
    pointer = _read_json(_pointer_path(root, session_id))
    item_id = pointer.get("item_id")
    ledger_path = _manifest_path(root, pointer.get("ledger_path"))
    if not isinstance(item_id, str) or not item_id or ledger_path is None or not ledger_path.is_file():
        return None

    doc = _read_toml(ledger_path)
    for item in doc.get("items", []):
        if item.get("id") == item_id and item.get("status") == "in_flight":
            return {
                "id": item_id,
                "stem": ledger_path.stem,
                "title": str(item.get("title", ""))[:TITLE_LIMIT],
            }
    return None


def _payload(items: list[dict[str, str]]) -> str:
    lines = [LINE_ONE]
    for item in items[:MAX_ITEMS]:
        lines.append(f"- {item['id']} [{item['stem']}] {item['title']}")
    if len(items) > MAX_ITEMS:
        lines.append(f"… +{len(items) - MAX_ITEMS} more")
    lines.append(GUIDANCE)
    payload = "\n".join(lines)
    if len(payload) > MAX_PAYLOAD_CHARS:
        return payload[:MAX_PAYLOAD_CHARS]
    return payload


def applies(data: dict) -> bool:
    source = data.get("source")
    return source is None or source in SESSIONSTART_SOURCES


def run(data: dict) -> Optional[HookResult]:
    root = find_project_root(data.get("cwd"))
    if root is None:
        return None

    pointer_item = _pointer_item(root, data.get("session_id"))
    if pointer_item is not None:
        return HookResult(kind="inject", payload=_payload([pointer_item]), module_name=MODULE_NAME)

    items = _campaign_inflight_items(root / "dev" / "campaigns")
    if not items:
        return None
    return HookResult(kind="inject", payload=_payload(items), module_name=MODULE_NAME)

File: .claude/hooks/modules/sessionstart/12_inflight_reanchor.ts
/** §in-flight re-anchor — a seat that claimed a row gets that row back at SessionStart.
 *
 *  The pointer is `.claude/state/ledger-active-<session_id>.json`, written by the ledger CLI's
 *  `claim`/`focus`. When it names a row that is still in_flight, this injects that one row and the
 *  command that prints its brief (laws included). A seat with no pointer gets nothing: listing every
 *  seat's in-flight rows to every session was noise, and was removed with the Stop-time in-flight
 *  warnings and the SessionStart law dump (hook audit, 2026-09-22).
 *
 *  The guidance text is READ AS AN INSTRUCTION, so it must name a command that exists.
 */
import { existsSync, readFileSync, statSync } from "node:fs";
import { join, relative } from "node:path";

import { HookResult, hookResult } from "../../orchestrator/result";
import { findProjectRoot } from "../../lib/project_root";

export const MODULE_NAME = "12_inflight_reanchor";
const SESSIONSTART_SOURCES: ReadonlySet<string> = new Set(["startup", "resume", "clear", "compact"]);
const MAX_PAYLOAD_CHARS = 500;
const TITLE_LIMIT = 60;

type HookEvent = Record<string, unknown>;

interface Item {
  id: string;
  ledger: string;
  title: string;
}

const safeId = (value: unknown): string =>
  String(value ?? "default").replace(/[^a-zA-Z0-9._-]/g, "_").slice(0, 128) || "default";

function readJson(path: string): Record<string, unknown> {
  try {
    const data = JSON.parse(readFileSync(path, "utf8")) as unknown;
    return data !== null && typeof data === "object" && !Array.isArray(data)
      ? (data as Record<string, unknown>)
      : {};
  } catch {
    return {};
  }
}

/** Returns null on a parse error — the caller treats an unreadable manifest as absent rather than
 *  as a finding, which is why this collapses the error instead of propagating. */
function readToml(path: string): Record<string, unknown> | null {
  try {
    const data = Bun.TOML.parse(readFileSync(path, "utf8")) as unknown;
    return data !== null && typeof data === "object" && !Array.isArray(data)
      ? (data as Record<string, unknown>)
      : null;
  } catch {
    return null;
  }
}

function pointerItem(root: string, sessionId: unknown): Item | null {
  const pointer = readJson(join(root, ".claude", "state", `ledger-active-${safeId(sessionId)}.json`));
  const itemId = pointer.item_id;
  const rawPath = pointer.ledger_path;
  if (typeof itemId !== "string" || !itemId || typeof rawPath !== "string" || !rawPath) return null;
  const ledgerPath = rawPath.startsWith("/") ? rawPath : join(root, rawPath);
  if (!existsSync(ledgerPath) || !statSync(ledgerPath).isFile()) return null;

  const doc = readToml(ledgerPath);
  if (doc === null) return null;
  const entries = Array.isArray(doc.items) ? (doc.items as Record<string, unknown>[]) : [];
  for (const item of entries) {
    if (item.id === itemId && item.status === "in_flight") {
      return { id: itemId, ledger: relative(root, ledgerPath), title: String(item.title ?? "").slice(0, TITLE_LIMIT) };
    }
  }
  return null;
}

function payload(item: Item): string {
  const text = [
    `§in-flight re-anchor — this session claimed ${item.id} [${item.ledger}] ${item.title}`,
    `Brief and laws: bun .dev/campaigns/manifest.ts ${item.ledger} packet ${item.id}`,
  ].join("\n");
  // A guard against a runaway title or path, not a text operation.
  return text.length > MAX_PAYLOAD_CHARS ? text.slice(0, MAX_PAYLOAD_CHARS) : text;
}

export function applies(data: HookEvent): boolean {
  const source = data.source;
  return source === undefined || source === null || SESSIONSTART_SOURCES.has(String(source));
}

export function run(data: HookEvent): HookResult | null {
  const root = findProjectRoot(data.cwd as string | undefined);
  if (root === null) return null;
  const item = pointerItem(root, data.session_id);
  return item === null ? null : hookResult("inject", payload(item), MODULE_NAME);
}
